src/module/data/pseudo-documents/pseudo-document.mjs

```diff
@@ -0,0 +1,272 @@
+const { DocumentIdField, StringField } = foundry.data.fields;
+
+/** @import { PseudoDocumentMetadata } from "../../_types" */
+
+export default class PseudoDocument extends foundry.abstract.DataModel {
+  /**
+   * Pseudo-document metadata.
+   * @type {PseudoDocumentMetadata}
+   */
+  static get metadata() {
+    return {
+      documentName: null,
+      embedded: {},
+    };
+  }
+
+  /* -------------------------------------------------- */
+
+  /**
+   * Registered sheets.
+   * @type {Map<string, PseudoDocumentSheet>}
+   */
+  static #sheets = new Map();
+
+  /* -------------------------------------------------- */
+
+  /** @inheritdoc */
+  static defineSchema() {
+    return {
+      _id: new DocumentIdField({ initial: () => foundry.utils.randomID() }),
+    };
+  }
+
+  /* -------------------------------------------------- */
+
+  /**
+   * The id of this pseudo-document.
+   * @type {string}
+   */
+  get id() {
+    return this._id;
+  }
+
+  /* -------------------------------------------------- */
+
+  /**
+   * The document name of this pseudo document.
+   * @type {string}
+   */
+  get documentName() {
+    return this.constructor.metadata.documentName;
+  }
+
+  /* -------------------------------------------------- */
+
+  /**
+   * The uuid of this document.
+   * @type {string}
+   */
+  get uuid() {
+    let parent = this.parent;
+    while (!(parent instanceof PseudoDocument) && !(parent instanceof foundry.abstract.Document)) parent = parent.parent;
+    return [parent.uuid, this.documentName, this.id].join(".");
+  }
+
+  /* -------------------------------------------------- */
+
+  /**
+   * The parent document of this pseudo-document.
+   * @type {Document}
+   */
+  get document() {
+    let parent = this;
+    while (!(parent instanceof foundry.abstract.Document)) parent = parent.parent;
+    return parent;
+  }
+
+  /* -------------------------------------------------- */
+
+  /**
+   * The property path to this pseudo document relative to its parent document.
+   * @type {string}
+   */
+  get fieldPath() {
+    const fp = this.schema.fieldPath;
+    let path = fp.slice(0, fp.lastIndexOf("element") - 1);
+
+    if (this.parent instanceof PseudoDocument) {
+      path = [this.parent.fieldPath, this.parent.id, path].join(".");
+    }
+
+    return path;
+  }
+
+  /* -------------------------------------------------- */
+
+  /**
+   * Reference to the sheet of this pseudo-document, registered in a static map.
+   * @type {PseudoDocumentSheet|null}
+   */
+  get sheet() {
```

Comment:
I'm likely gonna make a follow-up PR once the base sheet class is merged to put this logic entirely in the sheet class - rather than a method to "destroy" a sheet, the pseudo-document should I feel rather have a method to grab/create a new sheet, which is logic that should reside in the sheet class.

Then the sheet class can be responsible for destroying itself when closing as well.